
<!DOCTYPE doctype html>

<html class="no-js" lang="en">
<head>
<meta charset="utf-8"/>
<meta content="width=device-width,initial-scale=1" name="viewport"/>
<meta content="ie=edge" http-equiv="x-ua-compatible"/>
<meta content="Baltica - Integrated differential junction usage (DJU)" name="description"/>
<link href="https://github.com/dieterich-lab/Baltica/integration.html" rel="canonical"/>
<meta content="Thiago Britto-Borges" name="author"/>
<meta content="Copy to clipboard" name="lang:clipboard.copy"/>
<meta content="Copied to clipboard" name="lang:clipboard.copied"/>
<meta content="en" name="lang:search.language"/>
<meta content="True" name="lang:search.pipeline.stopwords"/>
<meta content="True" name="lang:search.pipeline.trimmer"/>
<meta content="No matching documents" name="lang:search.result.none"/>
<meta content="1 matching document" name="lang:search.result.one"/>
<meta content="# matching documents" name="lang:search.result.other"/>
<meta content="[\s\-]+" name="lang:search.tokenizer"/>
<link href="assets/images/favicon.png" rel="shortcut icon"/>
<meta content="mkdocs-1.0.4, mkdocs-material-4.6.3" name="generator"/>
<title>6 - DJU methods result integration - Baltica Documentation</title>
<link href="assets/stylesheets/application.adb8469c.css" rel="stylesheet"/>
<link href="assets/stylesheets/application-palette.a8b3c06d.css" rel="stylesheet"/>
<meta content="#757575" name="theme-color"/>
<script src="assets/javascripts/modernizr.86422ebf.js"></script>
<link crossorigin="" href="https://fonts.gstatic.com" rel="preconnect"/>
<link href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&amp;display=fallback" rel="stylesheet"/>
<style>body,input{font-family:"Roboto","Helvetica Neue",Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono","Courier New",Courier,monospace}</style>
<link href="assets/fonts/material-icons.css" rel="stylesheet"/>
<link href="extra.css" rel="stylesheet"/>
</head>
<body data-md-color-accent="blue-grey" data-md-color-primary="grey" dir="ltr">
<svg class="md-svg">
<defs>
<svg height="448" id="__github" viewbox="0 0 416 448" width="416" xmlns="http://www.w3.org/2000/svg"><path d="M160 304q0 10-3.125 20.5t-10.75 19T128 352t-18.125-8.5-10.75-19T96 304t3.125-20.5 10.75-19T128 256t18.125 8.5 10.75 19T160 304zm160 0q0 10-3.125 20.5t-10.75 19T288 352t-18.125-8.5-10.75-19T256 304t3.125-20.5 10.75-19T288 256t18.125 8.5 10.75 19T320 304zm40 0q0-30-17.25-51T296 232q-10.25 0-48.75 5.25Q229.5 240 208 240t-39.25-2.75Q130.75 232 120 232q-29.5 0-46.75 21T56 304q0 22 8 38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0 37.25-1.75t35-7.375 30.5-15 20.25-25.75T360 304zm56-44q0 51.75-15.25 82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5T212 416q-19.5 0-35.5-.75t-36.875-3.125-38.125-7.5-34.25-12.875T37 371.5t-21.5-28.75Q0 312 0 260q0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25 30.875Q171.5 96 212 96q37 0 70 8 26.25-20.5 46.75-30.25T376 64q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34 99.5z" fill="currentColor"></path></svg>
</defs>
</svg>
<input autocomplete="off" class="md-toggle" data-md-toggle="drawer" id="__drawer" type="checkbox"/>
<input autocomplete="off" class="md-toggle" data-md-toggle="search" id="__search" type="checkbox"/>
<label class="md-overlay" data-md-component="overlay" for="__drawer"></label>
<a class="md-skip" href="#parsing-the-results-of-the-method" tabindex="0">
        Skip to content
      </a>
<header class="md-header" data-md-component="header">
<nav class="md-header-nav md-grid">
<div class="md-flex">
<div class="md-flex__cell md-flex__cell--shrink">
<a aria-label="Baltica Documentation" class="md-header-nav__button md-logo" href="https://github.com/dieterich-lab/Baltica" title="Baltica Documentation">
<i class="md-icon"></i>
</a>
</div>
<div class="md-flex__cell md-flex__cell--shrink">
<label class="md-icon md-icon--menu md-header-nav__button" for="__drawer"></label>
</div>
<div class="md-flex__cell md-flex__cell--stretch">
<div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
<span class="md-header-nav__topic">
              Baltica Documentation
            </span>
<span class="md-header-nav__topic">
              
                6 - DJU methods result integration
              
            </span>
</div>
</div>
<div class="md-flex__cell md-flex__cell--shrink">
<label class="md-icon md-icon--search md-header-nav__button" for="__search"></label>
<div class="md-search" data-md-component="search" role="dialog">
<label class="md-search__overlay" for="__search"></label>
<div class="md-search__inner" role="search">
<form class="md-search__form" name="search">
<input aria-label="search" autocapitalize="off" autocomplete="off" autocorrect="off" class="md-search__input" data-md-component="query" data-md-state="active" name="query" placeholder="Search" spellcheck="false" type="text"/>
<label class="md-icon md-search__icon" for="__search"></label>
<button class="md-icon md-search__icon" data-md-component="reset" tabindex="-1" type="reset">
        
      </button>
</form>
<div class="md-search__output">
<div class="md-search__scrollwrap" data-md-scrollfix="">
<div class="md-search-result" data-md-component="result">
<div class="md-search-result__meta">
            Type to start searching
          </div>
<ol class="md-search-result__list"></ol>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="md-flex__cell md-flex__cell--shrink">
<div class="md-header-nav__source">
<a class="md-source" data-md-source="github" href="https://github.com/dieterich-lab/Baltica/" title="Go to repository">
<div class="md-source__icon">
<svg height="24" viewbox="0 0 24 24" width="24">
<use height="24" width="24" xlink:href="#__github"></use>
</svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
</div>
</div>
</nav>
</header>
<div class="md-container">
<main class="md-main" role="main">
<div class="md-main__inner md-grid" data-md-component="container">
<div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav class="md-nav md-nav--primary" data-md-level="0">
<label class="md-nav__title md-nav__title--site" for="__drawer">
<a class="md-nav__button md-logo" href="https://github.com/dieterich-lab/Baltica" title="Baltica Documentation">
<i class="md-icon"></i>
</a>
    Baltica Documentation
  </label>
<div class="md-nav__source">
<a class="md-source" data-md-source="github" href="https://github.com/dieterich-lab/Baltica/" title="Go to repository">
<div class="md-source__icon">
<svg height="24" viewbox="0 0 24 24" width="24">
<use height="24" width="24" xlink:href="#__github"></use>
</svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="README.html" title="1 - Index">
      1 - Index
    </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="intro.html" title="2 - Introduction">
      2 - Introduction
    </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="setup.html" title="3 - Getting started">
      3 - Getting started
    </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="workflows.html" title="4 - Methods">
      4 - Methods
    </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="benchmark.html" title="5 - Benchmark with artificial transcripts">
      5 - Benchmark with artificial transcripts
    </a>
</li>
<li class="md-nav__item md-nav__item--active">
<input class="md-toggle md-nav__toggle" data-md-toggle="toc" id="__toc" type="checkbox"/>
<label class="md-nav__link md-nav__link--active" for="__toc">
        6 - DJU methods result integration
      </label>
<a class="md-nav__link md-nav__link--active" href="integration.html" title="6 - DJU methods result integration">
      6 - DJU methods result integration
    </a>
<nav class="md-nav md-nav--secondary">
<label class="md-nav__title" for="__toc">Table of contents</label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="#parsing-the-results-of-the-method">
    Parsing the results of the method
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#annotating-the-results">
    Annotating the results
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#selecting-optimal-parameters-for-de-novo-transcriptome-assembly">
    Selecting optimal parameters for de novo transcriptome assembly
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#assigning-as-type">
    Assigning AS type
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#simplify-the-as-event">
    Simplify the AS event
  </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="release-notes.html" title="Release notes">
      Release notes
    </a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav class="md-nav md-nav--secondary">
<label class="md-nav__title" for="__toc">Table of contents</label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="#parsing-the-results-of-the-method">
    Parsing the results of the method
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#annotating-the-results">
    Annotating the results
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#selecting-optimal-parameters-for-de-novo-transcriptome-assembly">
    Selecting optimal parameters for de novo transcriptome assembly
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#assigning-as-type">
    Assigning AS type
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#simplify-the-as-event">
    Simplify the AS event
  </a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-content">
<article class="md-content__inner md-typeset"><a class="md-content__icon pdf-download-btn" download href="pdf/baltica_docs.pdf" title="Download"><i class="fa fas fa-download"></i><small> PDF</small></a>
<a class="md-icon md-content__icon" href="https://github.com/dieterich-lab/Baltica/tree/master/docs/integration.md" title="Edit this page"></a>
<h1>6 - DJU methods result integration</h1>
<h2 id="parsing-the-results-of-the-method">Parsing the results of the method<a class="headerlink" href="#parsing-the-results-of-the-method" title="Permanent link">¶</a></h2>
<p>The first step in the analysis workflow is parsing and processing the DJU methods' output with <code>scripts/parse_{method}_output.R</code> scripts as follows:</p>
<ol>
<li>The resulting text output from the DJU methods is parsed and loaded as R data frames</li>
<li>The data frames are pivoted in a longer format to have one junction and one comparison per row </li>
<li>Finally, SJ not called significant are discarded </li>
</ol>
<p>The default significant cut-off is an adjusted p-value &lt; 0.05 for JunctionSeq and Leafcutter or probability of changing &gt; 0.90 for Majiq.</p>
<p>The output results in <code>{method}/{method}_junction.csv</code> are in a <code>tidy</code> format and can be used for downstream analysis. </p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Although it is generally good to filter results with small effect sizes, discarding results with small deltaPSI can be problematic. First, it's not trivial to assign deltaPSI value to JunctionSeq results, since the tool does not detect splicing events. Second, SJ with small deltaPSI may indicate RNA degradation in a cell with intact degradation machinery. </p>
</div>
<h2 id="annotating-the-results">Annotating the results<a class="headerlink" href="#annotating-the-results" title="Permanent link">¶</a></h2>
<p>We annotate the results with information from the gene and transcript hosting the SJ. For this, we use the <em>de novo</em> transcript annotation at <code>stringtie/merged/merged.combined.gtf</code>. It's common the multiple transcripts share an intron, and so a single intron may be annotated with multiple transcripts.</p>
<p>One challenge for the integration of DJU methods' results is the different genomic coordinates system. The coordinates system's differences are due to the method implementation design: methods can be 0-indexed (BED format) versus 1-indexed (GTF format) or use the exonic versus intronic coordinates to represent the SJ genomic position. <br/>
We propose a <code>filter_hits_by_diff</code> (below) that discard any hits with more than two bp differences from overlapping features between the reference (query) and the SJ (subject). Using introns obtained in that annotation as a reference, and <code>filter_hits_by_diff</code>, Baltica enables the reconciliation of the multiple DJU results.</p>
<div class="codehilite"><pre><span></span><code><span class="n">filter_hits_by_diff</span> <span class="o">&lt;-</span> <span class="nf">function</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">subject</span><span class="p">,</span> <span class="n">max_start</span> <span class="o">=</span> <span class="m">2</span><span class="p">,</span> <span class="n">max_end</span> <span class="o">=</span> <span class="m">2</span><span class="p">)</span> <span class="p">{</span>
  <span class="nf">stopifnot</span><span class="p">(</span><span class="nf">is</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="s">"GRanges"</span><span class="p">))</span>
  <span class="nf">stopifnot</span><span class="p">(</span><span class="nf">is</span><span class="p">(</span><span class="n">subject</span><span class="p">,</span> <span class="s">"GRanges"</span><span class="p">))</span>
  <span class="n">hits</span> <span class="o">&lt;-</span> <span class="nf">findOverlaps</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">subject</span><span class="p">)</span>
  <span class="n">query</span> <span class="o">&lt;-</span> <span class="n">query</span><span class="p">[</span><span class="nf">queryHits</span><span class="p">(</span><span class="n">hits</span><span class="p">)]</span>
  <span class="n">subject</span> <span class="o">&lt;-</span> <span class="n">subject</span><span class="p">[</span><span class="nf">subjectHits</span><span class="p">(</span><span class="n">hits</span><span class="p">)]</span>
  <span class="n">start_dif</span> <span class="o">&lt;-</span> <span class="nf">abs</span><span class="p">(</span><span class="nf">start</span><span class="p">(</span><span class="n">query</span><span class="p">)</span> <span class="o">-</span> <span class="nf">start</span><span class="p">(</span><span class="n">subject</span><span class="p">))</span>
  <span class="n">end_dif</span> <span class="o">&lt;-</span> <span class="nf">abs</span><span class="p">(</span><span class="nf">end</span><span class="p">(</span><span class="n">query</span><span class="p">)</span> <span class="o">-</span> <span class="nf">end</span><span class="p">(</span><span class="n">subject</span><span class="p">))</span>
  <span class="n">hits</span> <span class="o">&lt;-</span> <span class="n">hits</span><span class="p">[</span><span class="n">start_dif</span> <span class="o">&lt;=</span> <span class="n">max_start</span> <span class="o">&amp;</span> <span class="n">end_dif</span> <span class="o">&lt;=</span> <span class="n">max_end</span><span class="p">]</span>
  <span class="n">hits</span>
<span class="p">}</span>
</code></pre></div>
<p>These are the columns assigned after the annotation:</p>
<table>
<thead>
<tr>
<th>Column name</th>
<th>Description</th>
<th>Note</th>
</tr>
</thead>
<tbody>
<tr>
<td>comparison</td>
<td>pairwise comparison as <code>{case}_vs_{control}</code></td>
<td></td>
</tr>
<tr>
<td>chr</td>
<td>seqname or genomic contig</td>
<td></td>
</tr>
<tr>
<td>start</td>
<td>intron start position for the SJ</td>
<td>1-index</td>
</tr>
<tr>
<td>end</td>
<td>intron end position</td>
<td>1-index</td>
</tr>
<tr>
<td>strand</td>
<td>RNA strand that encodes that gene</td>
<td></td>
</tr>
<tr>
<td>gene</td>
<td>the gene symbol</td>
<td></td>
</tr>
<tr>
<td>e2</td>
<td>acceptor exon number, if in + strand otherwise donor exon</td>
<td></td>
</tr>
<tr>
<td>e1</td>
<td>donor exon number, if in the + strand otherwise acceptor exon</td>
<td></td>
</tr>
<tr>
<td>tx_id</td>
<td>transcript identifier from the combined annotation</td>
<td></td>
</tr>
<tr>
<td>transcript_name</td>
<td>transcript name</td>
<td></td>
</tr>
<tr>
<td>class_code</td>
<td>association between reference transcript and novel transcript</td>
<td><a href="https://doi.org/10.12688/f1000research.23297.1">see fig 1 from this paper for details</a></td>
</tr>
</tbody>
</table>
<p><strong>Table 6.1: Columns added after annotation </strong></p>
<h2 id="selecting-optimal-parameters-for-de-novo-transcriptome-assembly">Selecting optimal parameters for <em>de novo</em> transcriptome assembly<a class="headerlink" href="#selecting-optimal-parameters-for-de-novo-transcriptome-assembly" title="Permanent link">¶</a></h2>
<p>We found that the parameters used to obtain the <em>de novo</em> transcriptome are critical for maximum integration between the GTF and the SJ from DJU methods. <strong>Fig 6.1</strong> shows a parameter scan where we vary the group, <code>-j</code> (minimum junction coverage), <code>-c</code> (minimum coverage), and <code>-f</code> (minimum isoform proportion) and compute the number of transcripts that match with SJ called significant. As expected, the merged annotation and not the group-specific annotation have the highest rate of annotated introns. The crucial result here is the dependency of the <code>-f</code> parameter, which is also associated with an increased number of annotated introns. As we confirmed this behavior in other datasets, we decided to use <code>-c 3 -j 3 -f 0.01</code> as default values in Baltica. The higher coverage (<code>-c</code> and <code>-j</code>) values counter the potential noise of transcripts with low abundance.</p>
<p><img alt="" src="img/stringtie_parameter_scan_heatmap.png"/> </p>
<p><strong> Fig. 6.1: Parameters scan to maximize the number of introns annotated </strong> We have run Stringtie with multiple for group annotations and merged annotation. We use a series of parameters: junction coverage of 1, 2, or 3; coverage of 1, 2, 3, and minimum isoform fraction of 0.1, 0.01, or 0.001.  </p>
<h2 id="assigning-as-type">Assigning AS type<a class="headerlink" href="#assigning-as-type" title="Permanent link">¶</a></h2>
<p>Identifying the type of AS is critical to understand a potential molecular mechanism for AS events. Many factors influence splicing. Among then, splicing factors probably the most well-studied and can have a specific function in splicing regulation. <a href="https://www.uniprot.org/uniprot/Q01130">SRSF2</a> is a relevant example in this context. SRSF2 is splicing factors from the SR family that are known for auto-regulation. In certain conditions, the SRSF2 transcript can activate the nonsense-mediated decay by either including a new exon containing a premature stop codon or an intron in it's 3' UTR. These changes lead to the transcript degradation and overall reduction of the gene expression. SRSF2 increased the probability of exons with weak splice site signals to be included in the transcript. Thus, the reduction of SRSF2 protein level leads to widespread exon skipping. Identifying such patterns is critical to understanding which splicing regulators are driving the observed splicing changes.</p>
<p>In Baltica, we use a geometric approach to define AS in three classes:
- ES, for exon skipping
- A3SS, for alternative 3' splice-site
- A5SS, for alternative 5' splice-site</p>
<p>Figure 6.2 details how we use the distance between features start and end to determine the AS type.</p>
<p><img alt="" src="img/Baltica_as_type.png"/> </p>
<p><strong> Fig. 6.2: AS type assignment in Baltica. </strong> Baltica uses the genomic coordinates from the SJ and its overlapping exons to assigning AS type to SJ and it's overlapping exons. Because many exons may be affected, multiple assignments are output. Donor and acceptor exons are assigned as JS and JE, respectively. </p>
<h2 id="simplify-the-as-event">Simplify the AS event<a class="headerlink" href="#simplify-the-as-event" title="Permanent link">¶</a></h2>
<p>Because most of the final users are only interested in the list of genomic ranges, gene names, or event types, we offer a simplified output that removes the excess of redundant information. This step is useful for generating a final report. </p>
<p>Below the <code>{r} sessionInfo()</code> output for packages loaded for the analysis workflow.</p>
<div class="codehilite"><pre><span></span><code>R version 3.6.0 (2019-04-26)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Debian GNU/Linux 8 (jessie)

Matrix products: default
BLAS:   /beegfs/biosw/R/3.6.0/lib/R/lib/libRblas.so
LAPACK: /beegfs/biosw/R/3.6.0/lib/R/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C
 [9] LC_ADDRESS=C               LC_TELEPHONE=C
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets
[8] methods   base

other attached packages:
 [1] openxlsx_4.1.5       reshape2_1.4.4       rtracklayer_1.46.0
 [4] GenomicRanges_1.38.0 GenomeInfoDb_1.22.1  IRanges_2.20.2
 [7] S4Vectors_0.24.4     BiocGenerics_0.32.0  optparse_1.6.6
[10] dplyr_0.8.5          readr_1.3.1          stringr_1.4.0
[13] tidyr_1.0.3

loaded via a namespace (and not attached):
 [1] zip_2.0.4                   Rcpp_1.0.4.6
 [3] plyr_1.8.6                  pillar_1.4.3
 [5] compiler_3.6.0              XVector_0.26.0
 [7] bitops_1.0-6                tools_3.6.0
 [9] zlibbioc_1.32.0             lattice_0.20-38
[11] lifecycle_0.2.0             tibble_3.0.1
[13] pkgconfig_2.0.3             rlang_0.4.6
[15] Matrix_1.2-17               DelayedArray_0.12.3
[17] cli_2.0.2                   GenomeInfoDbData_1.2.2
[19] Biostrings_2.54.0           vctrs_0.2.4
[21] hms_0.5.3                   grid_3.6.0
[23] getopt_1.20.3               tidyselect_1.0.0
[25] Biobase_2.46.0              glue_1.4.0
[27] R6_2.4.1                    fansi_0.4.1
[29] BiocParallel_1.20.1         XML_3.99-0.3
[31] purrr_0.3.4                 magrittr_1.5
[33] matrixStats_0.56.0          GenomicAlignments_1.22.1
[35] Rsamtools_2.2.3             ellipsis_0.3.0
[37] SummarizedExperiment_1.16.1 assertthat_0.2.1
[39] stringi_1.4.6               RCurl_1.98-1.2
[41] crayon_1.3.4
</code></pre></div>
</article>
</div>
</div>
</main>
<footer class="md-footer">
<div class="md-footer-nav">
<nav class="md-footer-nav__inner md-grid">
<a class="md-flex md-footer-nav__link md-footer-nav__link--prev" href="benchmark.html" rel="prev" title="5 - Benchmark with artificial transcripts">
<div class="md-flex__cell md-flex__cell--shrink">
<i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
</div>
<div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
<span class="md-flex__ellipsis">
<span class="md-footer-nav__direction">
                  Previous
                </span>
                5 - Benchmark with artificial transcripts
              </span>
</div>
</a>
<a class="md-flex md-footer-nav__link md-footer-nav__link--next" href="release-notes.html" rel="next" title="Release notes">
<div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
<span class="md-flex__ellipsis">
<span class="md-footer-nav__direction">
                  Next
                </span>
                Release notes
              </span>
</div>
<div class="md-flex__cell md-flex__cell--shrink">
<i class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
</div>
</a>
</nav>
</div>
<div class="md-footer-meta md-typeset">
<div class="md-footer-meta__inner md-grid">
<div class="md-footer-copyright">
<div class="md-footer-copyright__highlight">
            Thiago Britto-Borges, Dieterichlab - MIT License
          </div>
        
        powered by
        <a href="https://www.mkdocs.org" rel="noopener" target="_blank">MkDocs</a>
        and
        <a href="https://squidfunk.github.io/mkdocs-material/" rel="noopener" target="_blank">
          Material for MkDocs</a>
</div>
</div>
</div>
</footer>
</div>
<script src="assets/javascripts/application.c33a9706.js"></script>
<script>app.initialize({version:"1.0.4",url:{base:"."}})</script>
</body>
</html>