
<!DOCTYPE html>

<html class="no-js" lang="en">
<head>
<meta charset="utf-8"/>
<meta content="width=device-width,initial-scale=1" name="viewport"/>
<meta content="Baltica - Integrated differential junction usage (DJU)" name="description"/>
<meta content="Thiago Britto-Borges" name="author"/>
<link href="https://github.com/dieterich-lab/Baltica/workflows.html" rel="canonical"/>
<link href="assets/images/favicon.png" rel="icon"/>
<meta content="mkdocs-1.2.2, mkdocs-material-7.2.0" name="generator"/>
<title>Workflow implementation - Baltica Documentation</title>
<link href="assets/stylesheets/main.fe914879.min.css" rel="stylesheet"/>
<link href="assets/stylesheets/palette.ba0d045b.min.css" rel="stylesheet"/>
<link crossorigin="" href="https://fonts.gstatic.com" rel="preconnect"/>
<link href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&amp;display=fallback" rel="stylesheet"/>
<style>:root{--md-text-font-family:"Roboto";--md-code-font-family:"Roboto Mono"}</style>
<link href="extra.css" rel="stylesheet"/>
</head>
<body data-md-color-accent="" data-md-color-primary="" data-md-color-scheme="slate" dir="ltr">
<script>function __prefix(e){return new URL(".",location).pathname+"."+e}function __get(e,t=localStorage){return JSON.parse(t.getItem(__prefix(e)))}</script>
<input autocomplete="off" class="md-toggle" data-md-toggle="drawer" id="__drawer" type="checkbox"/>
<input autocomplete="off" class="md-toggle" data-md-toggle="search" id="__search" type="checkbox"/>
<label class="md-overlay" for="__drawer"></label>
<div data-md-component="skip">
<a class="md-skip" href="#workflow-implementation">
          Skip to content
        </a>
</div>
<div data-md-component="announce">
</div>
<header class="md-header" data-md-component="header">
<nav aria-label="Header" class="md-header__inner md-grid">
<a aria-label="Baltica Documentation" class="md-header__button md-logo" data-md-component="logo" href="index.html" title="Baltica Documentation">
<img alt="logo" src="img/baltica_logo.ico"/>
</a>
<label class="md-header__button md-icon" for="__drawer">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"></path></svg>
</label>
<div class="md-header__title" data-md-component="header-title">
<div class="md-header__ellipsis">
<div class="md-header__topic">
<span class="md-ellipsis">
            Baltica Documentation
          </span>
</div>
<div class="md-header__topic" data-md-component="header-topic">
<span class="md-ellipsis">
            
              Workflow implementation
            
          </span>
</div>
</div>
</div>
<label class="md-header__button md-icon" for="__search">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"></path></svg>
</label>
<div class="md-search" data-md-component="search" role="dialog">
<label class="md-search__overlay" for="__search"></label>
<div class="md-search__inner" role="search">
<form class="md-search__form" name="search">
<input aria-label="Search" autocapitalize="off" autocomplete="off" autocorrect="off" class="md-search__input" data-md-component="search-query" name="query" placeholder="Search" required="" spellcheck="false" type="text"/>
<label class="md-search__icon md-icon" for="__search">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"></path></svg>
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"></path></svg>
</label>
<nav aria-label="Search" class="md-search__options">
<button aria-label="Clear" class="md-search__icon md-icon" tabindex="-1" type="reset">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"></path></svg>
</button>
</nav>
</form>
<div class="md-search__output">
<div class="md-search__scrollwrap" data-md-scrollfix="">
<div class="md-search-result" data-md-component="search-result">
<div class="md-search-result__meta">
            Initializing search
          </div>
<ol class="md-search-result__list"></ol>
</div>
</div>
</div>
</div>
</div>
<div class="md-header__source">
<a class="md-source" data-md-component="source" href="https://github.com/dieterich-lab/Baltica/" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
</nav>
</header>
<div class="md-container" data-md-component="container">
<main class="md-main" data-md-component="main">
<div class="md-main__inner md-grid">
<div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Navigation" class="md-nav md-nav--primary" data-md-level="0">
<label class="md-nav__title" for="__drawer">
<a aria-label="Baltica Documentation" class="md-nav__button md-logo" data-md-component="logo" href="index.html" title="Baltica Documentation">
<img alt="logo" src="img/baltica_logo.ico"/>
</a>
    Baltica Documentation
  </label>
<div class="md-nav__source">
<a class="md-source" data-md-component="source" href="https://github.com/dieterich-lab/Baltica/" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="index.html">
        Index
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="intro.html">
        Introduction
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="setup.html">
        Getting started
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="proj-config.html">
        Project configuration
      </a>
</li>
<li class="md-nav__item md-nav__item--active">
<input class="md-nav__toggle md-toggle" data-md-toggle="toc" id="__toc" type="checkbox"/>
<label class="md-nav__link md-nav__link--active" for="__toc">
          Workflow implementation
          <span class="md-nav__icon md-icon"></span>
</label>
<a class="md-nav__link md-nav__link--active" href="workflows.html">
        Workflow implementation
      </a>
<nav aria-label="Table of contents" class="md-nav md-nav--secondary">
<label class="md-nav__title" for="__toc">
<span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
<ul class="md-nav__list" data-md-component="toc" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="#method-inclusion-criteria">
    Method inclusion criteria
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#quality-control-workflow">
    Quality control workflow
  </a>
<nav aria-label="Quality control workflow" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#software-dependencies">
    Software dependencies
  </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#dju-workflow">
    DJU workflow
  </a>
<nav aria-label="DJU workflow" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#leafcutter-workflow">
    Leafcutter workflow
  </a>
<nav aria-label="Leafcutter workflow" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#output">
    Output
  </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#majiq-workflow">
    Majiq workflow
  </a>
<nav aria-label="Majiq workflow" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#software-dependencies_1">
    Software dependencies
  </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#output_1">
    Output
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#junctionseq-workflow">
    JunctionSeq workflow
  </a>
<nav aria-label="JunctionSeq workflow" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#software-dependencies_2">
    Software dependencies
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#output_2">
    Output
  </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#stringtie-workflow">
    Stringtie workflow
  </a>
<nav aria-label="Stringtie workflow" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#software-dependencies_3">
    Software dependencies
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#parameters">
    Parameters
  </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="integration.html">
        DJU methods result integration
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="dev_guide.html">
        Development guidelines
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="faq.html">
        Frequently asked questions
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="release-notes.html">
        Release notes
      </a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Table of contents" class="md-nav md-nav--secondary">
<label class="md-nav__title" for="__toc">
<span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
<ul class="md-nav__list" data-md-component="toc" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="#method-inclusion-criteria">
    Method inclusion criteria
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#quality-control-workflow">
    Quality control workflow
  </a>
<nav aria-label="Quality control workflow" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#software-dependencies">
    Software dependencies
  </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#dju-workflow">
    DJU workflow
  </a>
<nav aria-label="DJU workflow" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#leafcutter-workflow">
    Leafcutter workflow
  </a>
<nav aria-label="Leafcutter workflow" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#output">
    Output
  </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#majiq-workflow">
    Majiq workflow
  </a>
<nav aria-label="Majiq workflow" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#software-dependencies_1">
    Software dependencies
  </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#output_1">
    Output
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#junctionseq-workflow">
    JunctionSeq workflow
  </a>
<nav aria-label="JunctionSeq workflow" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#software-dependencies_2">
    Software dependencies
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#output_2">
    Output
  </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#stringtie-workflow">
    Stringtie workflow
  </a>
<nav aria-label="Stringtie workflow" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#software-dependencies_3">
    Software dependencies
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#parameters">
    Parameters
  </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-content" data-md-component="content">
<article class="md-content__inner md-typeset"><a class="md-content__icon pdf-download-btn" download href="pdf/baltica_docs.pdf" title="Download"><i class="fa fas fa-download"></i><small> PDF</small></a>
<a class="md-content__button md-icon" href="https://github.com/dieterich-lab/Baltica/tree/master/docs/workflows.md" title="Edit this page">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"></path></svg>
</a>
<h1 id="workflow-implementation">Workflow implementation<a class="headerlink" href="#workflow-implementation" title="Permanent link">¶</a></h1>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This chapter is partially outdated.</p>
</div>
<p>This chapter details the implementation and usage of each workflow in Baltica.</p>
<p>Baltica comprises a collection of Snakemake workflows (in the SMK format). Each file determines a series of sub-tasks (rules). The sub-tasks run in a specific order; once the output of every rule is complete, the workflow is considered successful. We implemented the workflows following instructions and parameters suggested by the methods authors unless otherwise noted.</p>
<p><img alt="" class="center" src="img/Baltica_overview.png">
<strong>Fig. 1 - Baltica overview</strong>: As input (1), Baltica takes the alignment files in the BAM format, and transcriptome annotation and a configuration file that matches the sample names to the alignment file. The file also holds any workflow parameters. In the first optional step (2), Baltica produces a quality control report with MultiQC, which summarizes the results from FastQC and RSeQC. Next (3), Baltica computes the DJU methods and produces a <em>de novo</em> transcriptome with Stringtie. The novel exons and transcripts are indispensable for the integration step (4). Finally, the framework parses and integrates the output of the DJU methods.</img></p>
<p>Which are achieved by successively calling: </p>
<ol>
<li>Input</li>
<li>Quality control:   </li>
</ol>
<div class="codehilite"><pre><span></span><code>baltica qc &lt;config&gt;
</code></pre></div>
<ol>
<li>Differential Junction Usage (DJU) and <em>de novo</em> transcriptomics: <br/>
<code>baltica junctionseq &lt;config&gt;</code><br/>
<code>baltica majiq &lt;config&gt;</code><br/>
<code>baltica leafcutter &lt;config&gt;</code><br/>
<code>baltica stringtie &lt;config&gt;</code> </li>
<li>Integration: <br/>
<code>baltica analysis &lt;config&gt;</code></li>
</ol>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>The transcriptome annotation is critical, so make sure you use the same annotation during read alignment and Baltica.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The impact on read alignment on DJU methods results were not fully explored. You should expect different results with different read aligners. We recommend STAR. </p>
</div>
<h2 id="method-inclusion-criteria">Method inclusion criteria<a class="headerlink" href="#method-inclusion-criteria" title="Permanent link">¶</a></h2>
<p>There are a plethora of DJU methods, defined as methods that model SJ to identify differential splicing. Bellow is the inclusion criteria we used for Baltica. Methods are required to:</p>
<ul>
<li>use as input RNA-Seq read alignment in the BAM format</li>
<li>detect AS splicing as changes on SJs level, not at the transcript level</li>
<li>provide test statistic that compares a control group to a case group</li>
<li>output effect size estimates, such as the deltaPSI</li>
<li>detect unannotated SJ</li>
</ul>
<p>We are aware of other methods, such as Suppa <sup id="fnref:1"><a class="footnote-ref" href="#fn:1">1</a></sup> and rMATs <sup id="fnref:2"><a class="footnote-ref" href="#fn:2">2</a></sup>, also fit these criteria and aim to expand the catalog of supported methods in the future.</p>
<h2 id="quality-control-workflow">Quality control workflow<a class="headerlink" href="#quality-control-workflow" title="Permanent link">¶</a></h2>
<p>The first step comprises the quality control of sequenced libraries and read alignments.
This step aims to determine the success of sequencing and alignment.
Baltica includes workflows for RSeQC <sup id="fnref:3"><a class="footnote-ref" href="#fn:3">3</a></sup> and FastQC. MultiQC <sup id="fnref:4"><a class="footnote-ref" href="#fn:4">4</a></sup> summarizes the output from both tools.
In addition to the quality control, the tests may suggest biological differences among conditions.
For example, RSeQC provides the proportion of reads per feature in the input annotation, which may suggest an enrichment of reads mapping to intronic regions, indicating either intron retention or accumulation of unspliced mRNA.
RSeQC also implements an SJ saturation diagnostic, which quantifies the abundance of known and novel SJ.
This metric relates to the sequencing depth. This diagnostic is done by sampling subsets of the dataset to identify which proportion of annotated and novel introns are observed in the sub-samples. 
In conclusion, the quality control step serves to identify potential problems with the RNA-Seq library alignment and, potentially, direct on further troubleshooting and downstream analysis.</p>
<h3 id="software-dependencies">Software dependencies<a class="headerlink" href="#software-dependencies" title="Permanent link">¶</a></h3>
<table>
<thead>
<tr>
<th>Name</th>
<th>Version</th>
</tr>
</thead>
<tbody>
<tr>
<td>RSeQC</td>
<td>2.6.4</td>
</tr>
<tr>
<td>FastQC</td>
<td>0.11.8</td>
</tr>
<tr>
<td>MultiQC</td>
<td>0.8</td>
</tr>
</tbody>
</table>
<!-- [^1]: https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/
[^2]: http://rseqc.sourceforge.net/#usage-information
[^3]: https://multiqc.info/docs/
 -->
<h2 id="dju-workflow">DJU workflow<a class="headerlink" href="#dju-workflow" title="Permanent link">¶</a></h2>
<p>In term of implementation, the DJU tools use the following steps:  </p>
<ol>
<li>Extracting split reads from the alignment file    </li>
<li>Defining which SJ or events should be tested  </li>
<li>Modeling the SJ/events abundance  </li>
</ol>
<p>Unfortunately, there are differences among the utilized tools that lead to results that are not trivial to compare. </p>
<h3 id="leafcutter-workflow">Leafcutter workflow<a class="headerlink" href="#leafcutter-workflow" title="Permanent link">¶</a></h3>
<p>Leafcutter uses a series of scrips to extract the split reads from the BAM files. This step was recently changed to use <a href="https://github.com/griffithlab/regtools">regtools</a> to speed up the process. Our test shows that this new step affects the workflow results, and so we have not implemented the change in Baltica.</p>
<ol>
<li>Extracting intron from the alignments files: reads with M and N cigar are extracted from the alignments, giving a
minimum read overhang</li>
<li>Intron clustering: introns with at least <code>minclureads</code> (default: 30) reads and up to <code>maxintronlen</code> kb (
 default: 100000) are clustered. The clustering procedure iteratively discards introns supported by less than 
 <code>mincluratio</code> reads within a cluster.</li>
<li>Differential splicing analysis: Leafcutter uses a Dirichlet-Multinominal model to model the usage (proportion) of a 
giving SJ within a cluster and compare this usage among conditions</li>
</ol>
<!-- !!! info

  If you annotation use non-canonical chromosome names, you may have to change Leafcutter clustering scrip
 -->
<div class="admonition info">
<p class="admonition-title">Info</p>
<p>By default, Leafcutter clustering does not use the read strandedness information. In Baltica, we override this parameter and use the strand information for clustering.</p>
</div>
<!-- #### Software dependencies

Name | Version    
-----|---------
python | 2.7 
R | 3.5 
samtools | 1.9 
 -->
<!-- #### Parameters

Rule | Name | Default | Note
-----|-----|---------|------
`bam2junc` | `use_strand` | TRUE | 
`intron_clustering` | `minclureads` | 30 |
`intron_clustering` | `maxintronlen` | 100000 |
`intron_clustering` | `mincluratio` | 0.001 |
`differential_splicing` | `checkchrom` | TRUE |  
`differential_splicing` | `min_samples_per_group`  | 2 |
`differential_splicing` | `min_samples_per_intron` | 2 | 
`differential_splicing` | `fdr` | 0.05 |
 -->
<h4 id="output">Output<a class="headerlink" href="#output" title="Permanent link">¶</a></h4>
<p>The relevant output files from Leafcutter have the <code>_cluster_significance.txt</code> and <code>_effect_sizes.txt</code> suffix, which are computed for each comparison.</p>
<p>Column description:</p>
<p><code>*_cluster_significance.txt</code>:
1. <code>cluster</code>: <code>{chromosome}:{intron_start}:{intron_end}</code>
1. <code>Status</code>: is this cluster testable?
1. <code>loglr</code>: the log-likelihood ratio between the null model and alternative 
1. <code>df</code>: degrees of freedom, equal to the number of introns in the cluster minus one (assuming two groups)
1. <code>p</code> unadjusted p-value dor the under the asymptotic Chi-squared distribution</p>
<p><code>*_effect_sizes.txt</code>:
1. <code>intron</code>: intron identifier on the format <code>chromosome:intron_start:intron_end:cluster_id</code>
1. <code>es</code>: effect size
1. <code>{cond_1}</code>: fitted junction usage in condition <code>cond_1</code>
1. <code>{cond_2}</code>: fitted junction usage in condition <code>cond_2</code>
1. <code>deltapsi</code>: difference between usage in the two conditions  </p>
<h2 id="majiq-workflow">Majiq workflow<a class="headerlink" href="#majiq-workflow" title="Permanent link">¶</a></h2>
<p>Majiq workflow is implemented as follows:</p>
<ol>
<li>Create a configuration file (<code>majiq/build.ini</code>)</li>
<li><strong>Majiq build</strong> generates the Splice Graph database with exons and SJ from the RNA-Seq experiment and the reference annotation</li>
<li><strong>Majiq deltapsi</strong>: - computes PSI and deltaPSI and tests the if the deltaPSI changes between comparisons are significant </li>
<li><strong>Voila tsv</strong>: filter and process the Majiq output</li>
</ol>
<p>Majiq also provides a visualization with the <code>voila view</code> that we find helpful. </p>
<h3 id="software-dependencies_1">Software dependencies<a class="headerlink" href="#software-dependencies_1" title="Permanent link">¶</a></h3>
<table>
<thead>
<tr>
<th>Name</th>
<th>Version</th>
</tr>
</thead>
<tbody>
<tr>
<td>python</td>
<td>3.6</td>
</tr>
<tr>
<td>htslib</td>
<td>1.9</td>
</tr>
</tbody>
</table>
<!-- ### Parameters

Rule | Name | Default | Note
-----|------|---------|------
create_ini | `assembly` | | name of the assembly on the UCSC genome browser
create_ini | `strandness` | reverse | RNA-Sequencing library type 
create_ini | `read_len` |  100 | maximum read length
voila tsv | `majiq_threshold` | 0.2 | DeltaPSI cutoff for probability calculation
 -->
<h2 id="output_1">Output<a class="headerlink" href="#output_1" title="Permanent link">¶</a></h2>
<p>Baltica parses the files <code>*_voila.tsv</code> (one per comparison). One can read regarding Majiq's output at <a href="https://biociphers.bitbucket.io/majiq/VOILA_tsv.html">Majiq's online documentation</a></p>
<!-- <!-- build |   |  -->
<!--  `--min-experiments`
    - Description:integer or proportion of the minimum number of experiments a LSV event is observed to be considered
  - Default: 0.5

   `--min-intronic-cov`
  - Description: Minimum number of coverage an intron needs to be tested for intron retention events
  - Default: 0.01

   `--min-denovo`
  - Description: Minimum number of reads at all positions in an LSV to consider a de novo junction 
  - Default: 2

    `--minreads`
    - Description: Minimum number of reads at all positions in an LSV to consider an LSV
  - Default: 3

  `--minpos`
  - Description: Minimum number of start positions with at least one read in an LSV to consider that the LSV "exist in the data".
  - Default: 2

    `--markstacks`
  - Description: p-value threshold (or negative number to disable) to mark stack positions 
  - Default: 1e-07

    `--k`
  - Description: Number of positions to sample per iteration.
  - Default: 50

    `--m`
  - Description: Number of sampling steps using on bootstrap
  - Default: 30
  - baltica: -->
<p>--&gt;</p>
<!-- - For Majiq deltapsi:  

    `--binsize`
  - Description: Number of bins for the PSI value distribution 
  - Default: 0.025

    `--prior-minreads`
    - Description: Minimum number of reads at all to included 
  - Default: 20

    `--prior-minnonzero`
  - Minimum number of positions for the best set.
  - Default: 10
 -->
<!--    `--non-changing-threshold`
    - Description: None 
  - Default: 0.05

   `--probability-threshold` 
  - Description: None
  - default: off
 -->
<h2 id="junctionseq-workflow">JunctionSeq workflow<a class="headerlink" href="#junctionseq-workflow" title="Permanent link">¶</a></h2>
<p>JunctionSeq <sup id="fnref:5"><a class="footnote-ref" href="#fn:5">5</a></sup> tests statistical significance over difference usage among exonic and intronic disjoint genomic bins. It takes as input read count matrix obtained with QoRTs <sup id="fnref:6"><a class="footnote-ref" href="#fn:6">6</a></sup>, for annotated SJ, novel SJ, and exons, so in fact, JunctionSeq fits both the DEU and DJU classifications. 
Bins selected are testable as modeled with generalized linear models, as described in DEXSeq <sup id="fnref:7"><a class="footnote-ref" href="#fn:7">7</a></sup>, but reporting a test statistic at the genomic feature (exon or junction) and gene level.
Different from other DJU methods, JunctionSeq does not group the SJ in AS events, and so it does not compute PSI events.
By default, SJ with p.adjust &lt; 0.05 are called significant.</p>
<h3 id="software-dependencies_2">Software dependencies<a class="headerlink" href="#software-dependencies_2" title="Permanent link">¶</a></h3>
<table>
<thead>
<tr>
<th>Name</th>
<th>Version</th>
</tr>
</thead>
<tbody>
<tr>
<td>R</td>
<td>3.6</td>
</tr>
<tr>
<td>qorts</td>
<td>1.1.8</td>
</tr>
</tbody>
</table>
<p>Qorts depends on Java. We currently use it with Java 11.0.6. JunctionSeq itself relies on a series of BioConductor packages.</p>
<!-- ### Parameters

Rule | Name | Default | Note
-----|------|---------|------
qc   | `strandness` | reverse | 
qc   | `read_len`   | 100 |
qc   | `is_single_end` | True | 
 -->
<h3 id="output_2">Output<a class="headerlink" href="#output_2" title="Permanent link">¶</a></h3>
<p>Baltica parses files with the <code>*_sigGenes.results.txt.gz</code> suffix (one per comparison). Detailed output information are in the <a href="https://github.com/hartleys/JunctionSeq/blob/13a323dda5fae2d7e74b82230824affb747d938d/JunctionSeq/vignettes/JunctionSeq.Rnw#L514">JunctionSeq Package User Manual</a></p>
<h2 id="stringtie-workflow">Stringtie workflow<a class="headerlink" href="#stringtie-workflow" title="Permanent link">¶</a></h2>
<p>Baltica uses splice graph information to reconcile the SJ coordinates among methods and to assign AS type. </p>
<p>We process <em>de novo</em> transcriptomic workflow with Stringtie [@pertea_2015]. First, we merge the alignment files from biological replicates. Next, we compute <em>de novo</em> annotation with Stringtie (v1.3.5) with  <code>-c 3</code>, <code>-j 3</code>, and <code>-f 0.01</code>. Finally, the merge the multiple annotation with <code>gffcompare -r {reference_annotation.gtf} -R -V</code>. Details on the parameter selection are in the <a href="integration.html">Integration chapter</a>. </p>
<h3 id="software-dependencies_3">Software dependencies<a class="headerlink" href="#software-dependencies_3" title="Permanent link">¶</a></h3>
<table>
<thead>
<tr>
<th>Name</th>
<th>Version</th>
</tr>
</thead>
<tbody>
<tr>
<td>Stringtie</td>
<td>1.3.5</td>
</tr>
</tbody>
</table>
<h3 id="parameters">Parameters<a class="headerlink" href="#parameters" title="Permanent link">¶</a></h3>
<table>
<thead>
<tr>
<th>Rule</th>
<th>Name</th>
<th>Default</th>
<th>Note</th>
</tr>
</thead>
<tbody>
<tr>
<td>denovo_transcriptomics</td>
<td><code>strandness</code></td>
<td>reverse</td>
<td></td>
</tr>
<tr>
<td>denovo_transcriptomics</td>
<td><code>min_isoform_proportion</code></td>
<td>0.001</td>
<td></td>
</tr>
<tr>
<td>denovo_transcriptomics</td>
<td><code>min_junct_coverage</code></td>
<td>3</td>
<td></td>
</tr>
<tr>
<td>denovo_transcriptomics</td>
<td><code>minimum_read_per_bp_coverage</code></td>
<td>3</td>
<td></td>
</tr>
</tbody>
</table>
<div class="footnote">
<hr/>
<ol>
<li id="fn:1">
<p>Juan L. Trincado, Juan C. Entizne, Gerald Hysenaj, Babita Singh, Miha Skalic, David J. Elliott, and Eduardo Eyras. SUPPA2: fast, accurate, and uncertainty-aware differential splicing analysis across multiple conditions. <em>Genome Biology</em>, March 2018. URL: <a href="https://doi.org/10.1186/s13059-018-1417-1">https://doi.org/10.1186/s13059-018-1417-1</a>, <a href="https://doi.org/10.1186/s13059-018-1417-1">doi:10.1186/s13059-018-1417-1</a>. <a class="footnote-backref" href="#fnref:1" title="Jump back to footnote 1 in the text">↩</a></p>
</li>
<li id="fn:2">
<p>Shihao Shen, Juw Won Park, Zhi-xiang Lu, Lan Lin, Michael D. Henry, Ying Nian Wu, Qing Zhou, and Yi Xing. Rmats: robust and flexible detection of differential alternative splicing from replicate rna-seq data. <em>Proceedings of the National Academy of Sciences</em>, 111(51):E5593–E5601, Dec 2014. URL: <a href="http://dx.doi.org/10.1073/pnas.1419161111">http://dx.doi.org/10.1073/pnas.1419161111</a>, <a href="https://doi.org/10.1073/pnas.1419161111">doi:10.1073/pnas.1419161111</a>. <a class="footnote-backref" href="#fnref:2" title="Jump back to footnote 2 in the text">↩</a></p>
</li>
<li id="fn:3">
<p>Liguo Wang, Shengqin Wang, and Wei Li. RSeQC: quality control of RNA-seq experiments. <em>Bioinformatics</em>, 28(16):2184–2185, June 2012. URL: <a href="https://doi.org/10.1093/bioinformatics/bts356">https://doi.org/10.1093/bioinformatics/bts356</a>, <a href="https://doi.org/10.1093/bioinformatics/bts356">doi:10.1093/bioinformatics/bts356</a>. <a class="footnote-backref" href="#fnref:3" title="Jump back to footnote 3 in the text">↩</a></p>
</li>
<li id="fn:4">
<p>Philip Ewels, Måns Magnusson, Sverker Lundin, and Max Käller. Multiqc: summarize analysis results for multiple tools and samples in a single report. <em>Bioinformatics</em>, 32(19):3047–3048, Jun 2016. URL: <a href="http://dx.doi.org/10.1093/bioinformatics/btw354">http://dx.doi.org/10.1093/bioinformatics/btw354</a>, <a href="https://doi.org/10.1093/bioinformatics/btw354">doi:10.1093/bioinformatics/btw354</a>. <a class="footnote-backref" href="#fnref:4" title="Jump back to footnote 4 in the text">↩</a></p>
</li>
<li id="fn:5">
<p>Stephen W. Hartley and James C. Mullikin. Detection and visualization of differential splicing in RNA-seq data with JunctionSeq. <em>Nucleic Acids Research</em>, pages gkw501, June 2016. URL: <a href="https://doi.org/10.1093/nar/gkw501">https://doi.org/10.1093/nar/gkw501</a>, <a href="https://doi.org/10.1093/nar/gkw501">doi:10.1093/nar/gkw501</a>. <a class="footnote-backref" href="#fnref:5" title="Jump back to footnote 5 in the text">↩</a></p>
</li>
<li id="fn:6">
<p>Stephen W. Hartley and James C. Mullikin. Qorts: a comprehensive toolset for quality control and data processing of rna-seq experiments. <em>BMC Bioinformatics</em>, Jul 2015. URL: <a href="http://dx.doi.org/10.1186/s12859-015-0670-5">http://dx.doi.org/10.1186/s12859-015-0670-5</a>, <a href="https://doi.org/10.1186/s12859-015-0670-5">doi:10.1186/s12859-015-0670-5</a>. <a class="footnote-backref" href="#fnref:6" title="Jump back to footnote 6 in the text">↩</a></p>
</li>
<li id="fn:7">
<p>S. Anders, A. Reyes, and W. Huber. Detecting differential usage of exons from RNA-seq data. <em>Genome Research</em>, 22(10):2008–2017, June 2012. URL: <a href="https://doi.org/10.1101/gr.133744.111">https://doi.org/10.1101/gr.133744.111</a>, <a href="https://doi.org/10.1101/gr.133744.111">doi:10.1101/gr.133744.111</a>. <a class="footnote-backref" href="#fnref:7" title="Jump back to footnote 7 in the text">↩</a></p>
</li>
</ol>
</div>
</article>
</div>
</div>
</main>
<footer class="md-footer">
<nav aria-label="Footer" class="md-footer__inner md-grid">
<a aria-label="Previous: Project configuration" class="md-footer__link md-footer__link--prev" href="proj-config.html" rel="prev">
<div class="md-footer__button md-icon">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"></path></svg>
</div>
<div class="md-footer__title">
<div class="md-ellipsis">
<span class="md-footer__direction">
                Previous
              </span>
              Project configuration
            </div>
</div>
</a>
<a aria-label="Next: DJU methods result integration" class="md-footer__link md-footer__link--next" href="integration.html" rel="next">
<div class="md-footer__title">
<div class="md-ellipsis">
<span class="md-footer__direction">
                Next
              </span>
              DJU methods result integration
            </div>
</div>
<div class="md-footer__button md-icon">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"></path></svg>
</div>
</a>
</nav>
<div class="md-footer-meta md-typeset">
<div class="md-footer-meta__inner md-grid">
<div class="md-footer-copyright">
<div class="md-footer-copyright__highlight">
            Thiago Britto-Borges, Dieterichlab - MIT License
          </div>
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" rel="noopener" target="_blank">
          Material for MkDocs
        </a>
</div>
</div>
</div>
</footer>
</div>
<div class="md-dialog" data-md-component="dialog">
<div class="md-dialog__inner md-typeset"></div>
</div>
<script id="__config" type="application/json">{"base": ".", "features": ["instant"], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "search": "assets/javascripts/workers/search.53c85856.min.js", "version": null}</script>
<script src="assets/javascripts/bundle.716f8af4.min.js"></script>
</body>
</html>