---
title: "Report on Baltica DJU method integration: *SRSF4*"
author: "Thiago Britto-Borges"
date: "`r Sys.Date()`"
output:
  rmdformats::html_clean
  
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
knitr::opts_knit$set(progress = FALSE, verbose = FALSE)
library(tidyverse)
library(knitr)
library(rmdformats)
library(scales)
library(ggrepel)
## Global options
format_power10 <- scales::trans_format("log10", scales::label_math())

opts_chunk$set(
  echo = FALSE,
  cache = FALSE,
  prompt = FALSE,
  tidy = FALSE,
  comment = NA,
  message = FALSE,
  warning = FALSE,
  error = TRUE
)
if (file.exists("baltica_report.RData")) {
  load("baltica_report.RData")
} else {
  message("Processing report data.")
  source("report_data.R")
  load("baltica_report.RData")
}
```

## Quality Control summary:
```{r}
multiqc_fastqc <- read_delim("qc/multiqc/multiqc_data/multiqc_fastqc.txt",
  col_types = cols(
    .default = col_character(),
    `Sequences flagged as poor quality` = col_double(),
    avg_sequence_length = col_double(),
    `Total Sequences` = col_double(),
    `%GC` = col_double(),
    total_deduplicated_percentage = col_double()
  ),
  delim = "\t"
)
```

<!-- MultiQC parses and summarizes the statistics from a several computational tools on the RNA-Seq workflow. It reports that  -->
FastQC reports that `r sum(multiqc_fastqc$adapter_content == "fail") ` out of `r nrow(multiqc_fastqc)` fail the adaptor content test. The average sequence length is `r mean(multiqc_fastqc$avg_sequence_length)` and the average number of reads per sample of `r sprintf("%4.2e", mean(multiqc_fastqc[['Total Sequences']]))`.
For more detail, see the accompanying MultiQC report.  
 
## Summary exon-exon junction reads  {.tabset .tabset-fade .tabset-pills}
Show plots related to exon-exon junction distribution on all condition:  

### Distribuition of total counts
Distribution of total (unique + multi-mapping) exon-exon junctions reads per sample.  


```{r plots_hist_total, fig.width = 10, fig.height = 7}

mean_ <- summarise(
  group_by(sj_out, sample),
  mean = mean(uniq_map_read_count + multi_map_read_count)
) %>%
  mutate(label = substr(as.character(format_power10(mean)), 1, 7))

sj_out %>%
  ggplot(aes(x = uniq_map_read_count + multi_map_read_count)) +
  geom_histogram(alpha = .3) +
  facet_wrap(~sample, ncol = 4) +
  geom_vline(data = mean_, aes(xintercept = mean), linetype = "longdash") +
  geom_text(data = mean_, aes(x = mean + 200, label = label, y = 55000), parse = T) +
  scale_x_log10(labels = trans_format("log10", label_math())) +
  xlab(expression(log[10](unique ~ reads + multimapping ~ reads))) +
  ylab("Frequency") +
  theme_bw(16)
```

### Stacked distribution of reads per annotation status 
Distribution of uniquely mapped reads per sample. Annotated and unannotated reads distributions are shown in different colors.  

```{r plots_hist_per_annotation, fig.width = 10, fig.height = 7}

summary_ <- sj_out %>%
  filter(uniq_map_read_count != 0) %>%
  group_by(sample, annotation) %>%
  summarise(mean = mean(uniq_map_read_count)) %>%
  mutate(label = substr(as.character(format_power10(mean)), 1, 7))

sj_out %>%
  filter(uniq_map_read_count != 0) %>%
  ggplot(aes(x = uniq_map_read_count, fill = annotation)) +
  geom_histogram(alpha = .6, bins = 50, position = "stack") +
  facet_wrap(~sample, ncol = 4) +
  geom_vline(data = summary_, aes(xintercept = mean, color = annotation), linetype = "longdash") +
  geom_text(data = summary_, aes(x = mean + 10, label = label, y = 55000), size = 3, parse = T) +
  scale_x_log10(labels = trans_format("log10", label_math())) +
  xlab(expression(log[10](unique ~ reads))) +
  ylab("Frequency") +
  theme_bw(16)
```

### Distribution of unannotated reads.
Distribution of uniquely mapped reads per sample, showing only unannotated reads.  

```{r plots_hist_unannotated, fig.width = 10, fig.height = 7, error=TRUE}
summary_ <- sj_out %>%
  filter(uniq_map_read_count != 0 & annotation == "unannotated") %>%
  group_by(sample) %>%
  summarise(
    mean = mean(uniq_map_read_count),
    annotation = first(annotation)
  ) %>%
  mutate(label = substr(as.character(format_power10(mean)), 1, 7))


sj_out %>%
  filter(uniq_map_read_count != 0 & annotation == "unannotated") %>%
  ggplot(aes(x = uniq_map_read_count, fill = annotation)) +
  geom_histogram(alpha = .6, bins = 50, position = "stack") +
  facet_wrap(~sample, ncol = 4) +
  geom_vline(data = summary_, aes(xintercept = mean), linetype = "longdash") +
  geom_text(data = summary_, aes(x = mean + 1, label = label, y = 650), size = 3, parse = T) +
  scale_x_log10(labels = trans_format("log10", label_math())) +
  xlab(expression(log[10](unique ~ reads))) +
  ylab("Frequency") +
  theme_bw(16)
```

### Distribution of reads by intron motif
Distribution of reads by intron motif.  
```{r plots_hist_intron_motif, fig.width = 10, fig.height = 7, echo=FALSE}
sj_out %>%
  filter(uniq_map_read_count != 0) %>%
  ggplot(aes(x = uniq_map_read_count, fill = intron_motif)) +
  facet_wrap(~sample, ncol = 4) +
  geom_histogram(alpha = .6, bins = 50, position = "stack") +
  scale_fill_brewer(palette = "Spectral") +
  scale_x_log10(labels = trans_format("log10", label_math())) +
  xlab(expression(log[10](unique ~ reads))) +
  ylab("Frequency") +
  theme_bw(16)
```

<!-- TODO: PCA -->
<!-- ### Principal componente analysis  -->
<!-- PCA of the 10k junction with highest variance -->


<!-- ```{r plot_pca, fig.width = 10, fig.height = 7, echo=FALSE} -->
<!-- plot_pca -->
<!-- ``` -->

## Leafcutter plots {.tabset .tabset-fade .tabset-pills}

### P-value distribution
```{r}
leafcutter_complete %>%
  ggplot(aes(x=p)) +
  labs(title="Leafcutter p.value distribution",
       x='p.value',
       y='Frequency') +
  geom_histogram( bins=50, alpha=0.5) + 
  facet_wrap(~comparison, nrow = 3) + 
  geom_vline(xintercept=0.05, color='red', linetype = "longdash") +
  theme_bw(base_size = 16) 
```

### P-adjust distribution
```{r}
leafcutter_complete %>%
  ggplot(aes(x=p.adjust)) +
  labs(title="Leafcutter p.adjust distribution",
       x='p.adjust',
       y='Frequency') +
  geom_histogram( bins=50, alpha=0.5) + 
  facet_wrap(~comparison, nrow = 3) + 
  geom_vline(xintercept=0.05, color='red', linetype = "longdash") +
  theme_bw(base_size = 16) 
```

### Volcano
```{r}
p <- leafcutter_complete %>% j
  ggplot(aes(deltapsi, -log10(p.adjust))) +
    geom_point(aes(colour=is_sig), alpha=.5, size=2) +
    geom_hline(yintercept = -log10(0.05), linetype='dashed', size=0.3) +
    geom_vline(xintercept = c(-0.1, 0.1), linetype='dashed', size=0.3) +
    scale_colour_manual(values = c("grey", "red")) +
    theme_minimal(base_size = 16) +
    labs(y=expression(-log[10](p.adjust)),
         x= expression(Delta~PSI),
         color=expression(abs(deltapsi) > 0.1 & p.adjust < 0.05),
         tittle='Leafcutter: volcano plot') +
    facet_wrap(~comparison, nrow = 3) 
p <- p +  geom_text_repel(
     data=leafcutter_complete[top_rows,],
     aes(label = genes),
     fontface = 'bold', color = 'firebrick',
     segment.color = 'grey50',
     size = 3,
     box.padding = unit(0.35, "lines"),
     point.padding = unit(0.3, "lines"),
     force = 5,
     nudge_y = 5,
     nudge_x = 5,
     max.iter = 10000)
p +   theme(legend.position="bottom")
```

## Results table
Click on coordinates to open a link to the genome browser.

```{r}
res %>%
  select(-c(transcript_name, class_code)) %>% 
  mutate(
    score = round(as.numeric(score), 4),
    comparison = as.factor(comparison),
    method = as.factor(method),
    gene_name = as.factor(gene_name),
    is_novel = as.factor(is_novel)
  ) %>%
  DT::datatable(.,
    filter = "top",
    rownames = FALSE,
    escape = FALSE,
    options = list(
      searchHighlight = TRUE,
      scrollX = TRUE,
      columnDefs = list(list(className = "dt-center"))
    )
  )
```


## Reproducibility {.tabset .tabset-fade .tabset-pills}

### Parameters used in the analysis:  
- Configuration file used by baltica
- Version and method citation table
- R session information

### Citations:  
```{r, echo=FALSE, collapse=T}
"TODO"
```

### Configuration:  
```{r, echo=FALSE, collapse=T}
pander::pander(config, compact = FALSE)
```

### Session information:    
```{r, echo=FALSE, collapse=T}
pander::pander(sessionInfo(), compact = FALSE)
```
