---
title: "Baltica report"
date: "`r Sys.Date()`"
# css: "custom.css"
output:
  rmdformats::downcute:
    self_contained: true
    lightbox: true
    gallery: false
params:
  rmd: baltica_report.Rmd
---


```{r libs, include = FALSE}
library(tidyverse)
library(knitr)
library(rmdformats)
library(scales)
library(ggrepel)
library(pander)

## Global options
format_power10 <- scales::trans_format("log10", scales::label_math())

opts_chunk$set(
  echo = FALSE,
  cache = FALSE,
  prompt = FALSE,
  tidy = FALSE,
  comment = NA,
  message = FALSE,
  warning = FALSE
)
# bibliography <- file.path(snakemake@scriptdir, "references.bib")
# is_fastqc <- F # file.exists(snakemake@params$fastqc)
# is_gffcompare <- file.exists((snakemake@params$gffcompare))
sj_annotated_file <- as.character(snakemake@input[[1]])
is_sj_annotate <- file.exists(sj_annotated_file)
leafcutter_file <- as.character(snakemake@input[[2]])
is_leafcutter <- file.exists(leafcutter_file)
config <- yaml::read_yaml(snakemake@params$config)


knitr::knit_hooks$set(
  error = function(x, options) {
    paste('\n\n<div class="alert alert-danger">',
      gsub("##", "\n", gsub("^##\ Error", "**Error**", x)),
      "</div>",
      sep = "\n"
    )
  },
  warning = function(x, options) {
    paste('\n\n<div class="alert alert-warning">',
      gsub("##", "\n", gsub("^##\ Warning:", "**Warning**", x)),
      "</div>",
      sep = "\n"
    )
  },
  message = function(x, options) {
    paste('\n\n<div class="alert alert-info">',
      gsub("##", "\n", x),
      "</div>",
      sep = "\n"
    )
  }
)
```

## Resources:

This is a report for the Baltica framework. Please see https://github.com/dieterich-lab/Baltica and https://dieterich-lab.github.io/Baltica/ for details.

## Summary on Leafcutter results: {.tabset .tabset-fade .tabset-pills}
```{r message=TRUE, warning=TRUE}
if (!is_leafcutter) {
  warning("Leafcutter result files not found.")
} else {
  message("See Leafcutter summary bellow.")
}
```

```{r, eval=is_leafcutter}

knitr::asis_output("### P-adjust distribution.\\n")

leafcutter_complete <- read_csv(leafcutter_file, col_types = cols(
  comparison = col_character(),
  intron = col_character(),
  logef = col_double(),
  ref_psi = col_double(),
  alt_psi = col_double(),
  deltapsi = col_double(),
  chr = col_character(),
  start = col_double(),
  end = col_double(),
  strand = col_character(),
  cluster = col_character(),
  status = col_character(),
  loglr = col_double(),
  df = col_double(),
  p = col_double(),
  p.adjust = col_double(),
  method = col_character()
))

leafcutter_complete <- leafcutter_complete %>%
  mutate(is_sig = abs(deltapsi) > 0.1 & p.adjust < 0.05)

leafcutter_complete %>%
  ggplot(aes(x = p)) +
  labs(
    title = "Leafcutter p.value distribution",
    x = "p.value",
    y = "Frequency"
  ) +
  geom_histogram(bins = 50, alpha = 0.5) +
  facet_wrap(~comparison, nrow = 3) +
  geom_vline(xintercept = 0.05, color = "red", linetype = "longdash") +
  theme_bw(base_size = 16)
```


```{r eval=is_leafcutter}
knitr::asis_output("### Volcano\\n")

p <- leafcutter_complete %>%
  ggplot(aes(deltapsi, -log10(p.adjust))) +
  geom_point(aes(colour = is_sig), alpha = .5, size = 2) +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", size = 0.3) +
  geom_vline(xintercept = c(-0.1, 0.1), linetype = "dashed", size = 0.3) +
  scale_colour_manual(values = c("grey", "red")) +
  theme_minimal(base_size = 16) +
  labs(
    y = expression(-log[10](p.adjust)),
    x = expression(Delta ~ PSI),
    color = "is_sig",
    tittle = "Leafcutter: volcano plot",
    caption = expression(
      paste("is_sig: |", Delta ~ PSI, "|> 0.1 & p.adj < 0.05")
    )
  ) +
  facet_wrap(~comparison, nrow = 3) +
  theme(legend.position = "bottom")
p
```

## Baltica table:
```{r warning=TRUE, message=TRUE}


if (!is_sj_annotate) {
  warning("Baltica table file not found.")
}
```


```{r echo=FALSE, eval=is_sj_annotate}
knitr::asis_output("### Common splice junctions \\n\\n")

df <- readr::read_csv(sj_annotated_file, col_types = cols(
  .default = col_double(),
  is_novel = col_logical(),
  gene_name = col_character(),
  transcript_name = col_character(),
  class_code = col_character(),
  exon_number = col_character(),
  coordinates = col_character(),
  as_type = col_character()
))

suppressPackageStartupMessages(library(ComplexHeatmap))
library(RColorBrewer)

methods <- c("rmats", "junctionseq", "majiq", "leafcutter", "orthogonal")

mat <- df %>%
  dplyr::select(matches("-vs-")) %>%
  as.matrix(.)

comparisons <- unique(
  gsub(x = colnames(mat), str_c(c(methods, "_"), collapse = "|"), "")
)

mat[mat > 0.95] <- 1
mat[mat < 0.95] <- 0
mat[is.na(mat)] <- 0
mat <- mat[rowSums(mat) >= 1, ]


comb_mat <- make_comb_mat(mat)

group <- str_extract(colnames(mat), str_c(comparisons, collapse = "|"))
method <- str_extract(colnames(mat), str_c(methods, collapse = "|"))

UpSet(
  comb_mat,
  comb_order = order(rev(comb_degree(comb_mat))),
  pt_size = unit(1.5, "mm"),
  lwd = 0.5,
  right_annotation = rowAnnotation(
    "Set size" = anno_barplot(
      set_size(comb_mat),
      border = FALSE,
      gp = gpar(fill = "black"),
      width = unit(2, "cm")
    ),
    # https://stackoverflow.com/a/47268265/1694714
    group = group,
    method = method,
    col = list(
      group = setNames(
        brewer.pal(n = length(comparisons), name = "Set2"), comparisons
      ),
      method = setNames(
        brewer.pal(n = length(methods), name = "Set1"), methods
      )
    )
  )
)
```



```{r echo=FALSE, eval=is_sj_annotate}
knitr::asis_output("### Baltica Table \\n
Click on coordinates to go to the UCSC Genome Browser; on the â–¼ to see
intron annotation. Multi columns are sortable by holding <kbd>SHIFT</kbd>.")

library(reactable)

# simplify <- function(x, remove = c("JS", "JE")) {
#   x %>%
#     str_split(";") %>%
#     lapply(., function(x) {
#       setdiff(x, remove)
#     }) %>%
#     lapply(., paste0, collapse = ";") %>%
#     unlist()
# }

df <- df %>%
  mutate_if(is.numeric, round, 3) %>%
  # mutate(as_type = simplify(as_type)) %>%
  dplyr::select(
    coordinates,
    gene_name,
    transcript_name,
    class_code,
    # as_type,
    exon_number,
    is_novel,
    everything()
  )

sticky_style <- list(
  position = "sticky",
  left = 0,
  background = "#fff",
  zIndex = 1,
  borderRight = "1px solid #eee"
)
comp_cols <- grepl("-vs-", colnames(df))
comp_cols <- colnames(df)[comp_cols]
comp_cols_mat <- str_split(comp_cols, "_", simplify = T)
comp_cols_mat <- as_tibble(comp_cols_mat)
comp_cols_mat["V3"] <- comp_cols
cols_groups <- split(comp_cols_mat$V3, comp_cols_mat$V2)
columnGroups <- lapply(
  names(cols_groups),
  function(x) colGroup(name = x, columns = cols_groups[[x]])
)
comp_cols_new_name <- lapply(seq_along(comp_cols), function(i) {
  colDef(
    name = sub("_[a-z0-9-]+", "", comp_cols[i]),
    cell = function(value) ifelse(is.na(value), "-", value)
  )
})
names(comp_cols_new_name) <- comp_cols
# from https://github.com/glin/reactable/issues/29#issuecomment-591752165
html <- function(x, inline = FALSE) {
  container <- if (inline) htmltools::span else htmltools::div
  container(dangerouslySetInnerHTML = list("__html" = x))
}


target_url <- "http://genome.ucsc.edu/cgi-bin/hgTracks?db=
{config$assembly}&position={new_value}"
# <p><strong>as_type: </strong>{as_type}</p>"

df %>%
  mutate(ex = str_glue_data(
    .,
    "<p><strong>transcript_name:</strong>{transcript_name}</p>\n
    <p><strong>class_code:</strong>{class_code}</p>\n
    <p><strong>exon_number:</strong>{exon_number}</p>"
  )) %>%
  mutate(ex = ifelse(is.na(gene_name), NA_character_, ex)) %>%
  dplyr::select(
    -transcript_name, -class_code, -exon_number
    # -as_type
  ) %>%
  reactable(
    bordered = TRUE, highlight = TRUE, compact = TRUE,
    height = 400,
    columnGroups = columnGroups,
    paginationType = "simple", defaultPageSize = 10,
    columns = modifyList(
      list(
        ex = colDef(show = FALSE),
        coordinates = colDef(
          cell = function(value, index) {
            # Render as a link
            new_value <- str_sub(value, 1, -3)
            url <- str_glue(target_url)
            htmltools::tags$a(
              href = url, target = "_blank", as.character(value)
            )
          },
          style = sticky_style,
          headerStyle = sticky_style,
          minWidth = 190
        ),
        gene_name = colDef(filterable = TRUE),
        is_novel = colDef(
          filterable = TRUE,
          cell = function(value) {
            case_when(
              is.na(value) ~ "-",
              value == TRUE ~ "\u2718",
              TRUE ~ "\u2713"
            )
          }
        )
      ), comp_cols_new_name
    ),
    details = function(index) {
      if (
        !is.na(.$ex[index])) {
        htmltools::tagList(html(.$ex[index]))
      }
    }
  )
```


## Reproducibility: {.tabset .tabset-fade .tabset-pills}

### Parameters used in the analysis:  
- Baltica configuration file
- R session information
- References for methods included in Baltica


### Configuration:  
```{r}
pander::pander(config, compact = FALSE)
```

### Session information:    
```{r}
pander::pander(sessionInfo(), compact = FALSE)
```

### References:  
